File: src/providers/tracer.py
import atexit
import json
import logging
import os
from datetime import datetime, timezone
from typing import Any, Optional

from providers.singleton import singleton


@singleton
class Tracer:
    """
    Tracer for LLM interactions. Writes one JSONL line per record, with a daily rotation.
    """

    def __init__(self, output_dir: str = "traces") -> None:
        """
        Initialize the Tracer.

        Parameters
        ----------
        output_dir : str
            Directory where traces will be saved (default: "traces").
        """
        self.output_dir = output_dir
        self._enabled = False
        self._current_date: Optional[str] = None
        self._file: Any = None
        self.generation: int = 0

        if self._enabled:
            os.makedirs(self.output_dir, exist_ok=True)

        atexit.register(self.stop)

    def enable(self) -> None:
        """
        Enable tracing and ensure the output directory exists.
        """
        self._enabled = True
        os.makedirs(self.output_dir, exist_ok=True)

    def disable(self) -> None:
        """
        Disable tracing and close any open file handle.
        """
        self._enabled = False
        self.stop()

    @property
    def enabled(self) -> bool:
        """
        Return whether tracing is currently enabled.
        """
        return self._enabled

    def gauge(self, llm_input: str, llm_output: list[dict[str, Any]]) -> None:
        """
        Record an LLM interaction with the given input and output.

        Parameters
        ----------
        llm_input : str
            The input prompt sent to the LLM.

        llm_output : list[dict[str, Any]]
            The output from the LLM, typically a list of message dictionaries.
        """
        if not self._enabled:
            return

        try:
            data = {
                "ts": datetime.now(timezone.utc).isoformat(),
                "generation": self.generation,
                "llm_input": llm_input,
                "llm_output": llm_output,
            }
            self._write(json.dumps(data, ensure_ascii=False))
        except Exception:
            logging.exception("tracer: failed to write record")

    def stop(self) -> None:
        """
        Close the current file handle.
        """
        if self._file:
            try:
                self._file.close()
            except Exception:
                pass

            self._file = None
            self._current_date = None

    def _write(self, line: str) -> None:
        """
        Write a line to the current day's JSONL file, rotating if the date has changed.

        Parameters
        ----------
        line : str
            The JSON string to write as a line in the file.
        """
        now = datetime.now(timezone.utc).strftime("%Y-%m-%d")

        if now != self._current_date:
            self.stop()
            filepath = os.path.join(self.output_dir, f"{now}.jsonl")
            self._file = open(filepath, "a", encoding="utf-8")
            self._current_date = now

        self._file.write(line + "\n")
        self._file.flush()

    def set_generation(self, generation: int) -> None:
        """
        Set the current generation number for tracing.

        Parameters
        ----------
        generation : int
            The generation number to set.
        """
        self.generation = generation

File: internal/tracer/tracer.go
package tracer

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/openmind/om1/internal/config"
	"github.com/openmind/om1/internal/logger"
	"github.com/openmind/om1/internal/tracer/tracetype"
	"go.uber.org/zap"
)

// TraceRecord is one LLM interaction, written as a JSONL line and handed to any subscriber.
type TraceRecord = tracetype.TraceRecord

// Tracer provides a simple mechanism to record LLM interactions.
type Tracer struct {
	mu          sync.Mutex
	outputDir   string
	enabled     bool
	currentDate string
	file        *os.File
	generation  int

	subscribers []chan<- TraceRecord

	qualityScoreCancel context.CancelFunc
	traceExportCancel  context.CancelFunc
}

var (
	tracerOnce     sync.Once
	tracerInstance *Tracer
)

// TracerProvider returns the singleton Tracer instance.
func TracerProvider() *Tracer {
	tracerOnce.Do(func() {
		tracerInstance = &Tracer{outputDir: "traces"}
	})
	return tracerInstance
}

// Start starts the tracer and its quality scorer if enabled in cfg, using a context that can be cancelled on shutdown.
func (t *Tracer) Start(ctx context.Context, cfg *config.TracerConfig, systemAPIKey string, log *zap.Logger) {
	if cfg == nil {
		return
	}

	if !cfg.Enabled {
		if cfg.QualityScorer != nil && cfg.QualityScorer.Enabled {
			log.Warn("tracer: quality_scorer.enabled is true but use_tracer.enabled is false -- quality scorer will not start")
		}
		if cfg.PrometheusExport != nil && cfg.PrometheusExport.Enabled {
			log.Warn("tracer: prometheus_export.enabled is true but use_tracer.enabled is false -- Prometheus trace export will not start")
		}
		return
	}
	t.Enable()

	t.startQualityScore(ctx, cfg.QualityScorer, systemAPIKey, log)
	t.startTraceExport(ctx, cfg.PrometheusExport, log)
}

// Enable turns tracing on and ensures the output directory exists.
func (t *Tracer) Enable() {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.enabled = true
	if err := os.MkdirAll(t.outputDir, 0o755); err != nil {
		logger.Get().Warn("tracer: failed to create output dir", zap.Error(err))
	}
}

// Disable turns tracing off and closes any open file handle.
func (t *Tracer) Disable() {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.enabled = false
	t.stopLocked()
}

// Enabled reports whether tracing is currently enabled.
func (t *Tracer) Enabled() bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.enabled
}

// SetGeneration sets the current generation number recorded with each trace.
func (t *Tracer) SetGeneration(generation int) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.generation = generation
}

// Subscribe returns a channel receiving every future TraceRecord; buffered and non-blocking, no Unsubscribe.
func (t *Tracer) Subscribe() <-chan TraceRecord {
	ch := make(chan TraceRecord, 32)
	t.mu.Lock()
	t.subscribers = append(t.subscribers, ch)
	t.mu.Unlock()
	return ch
}

// Gauge records an LLM interaction with the given input prompt and output.
func (t *Tracer) Gauge(llmInput string, llmOutput []map[string]any) {
	t.mu.Lock()

	if !t.enabled {
		t.mu.Unlock()
		return
	}

	if llmOutput == nil {
		llmOutput = []map[string]any{}
	}

	rec := TraceRecord{
		Timestamp:  time.Now().UTC().Format(time.RFC3339Nano),
		Generation: t.generation,
		LLMInput:   llmInput,
		LLMOutput:  llmOutput,
	}

	line, err := json.Marshal(rec)
	if err != nil {
		t.mu.Unlock()
		logger.Get().Warn("tracer: failed to marshal record", zap.Error(err))
		return
	}

	if err := t.writeLocked(line); err != nil {
		logger.Get().Warn("tracer: failed to write record", zap.Error(err))
	}

	subs := t.subscribers
	t.mu.Unlock()

	for _, ch := range subs {
		select {
		case ch <- rec:
		default:
			logger.Get().Warn("tracer: subscriber channel full, dropping trace record")
		}
	}
}

// Stop stops the tracer, its quality scorer, and its trace exporter, closing any open file handle.
func (t *Tracer) Stop() {
	t.stopQualityScore()
	t.stopTraceExport()

	t.mu.Lock()
	defer t.mu.Unlock()
	t.stopLocked()
}

// stopLocked closes the current file handle without acquiring the mutex.
func (t *Tracer) stopLocked() {
	if t.file != nil {
		_ = t.file.Close()
		t.file = nil
		t.currentDate = ""
	}
}

// writeLocked writes a line to the current trace file, rotating daily.
func (t *Tracer) writeLocked(line []byte) error {
	now := time.Now().UTC().Format("2006-01-02")

	if now != t.currentDate {
		t.stopLocked()
		path := filepath.Join(t.outputDir, "tracer_"+now+".jsonl")
		f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
		if err != nil {
			return err
		}
		t.file = f
		t.currentDate = now
	}

	if _, err := t.file.Write(append(line, '\n')); err != nil {
		return err
	}
	return t.file.Sync()
}
